### PR TITLE
Add maxNumberOfNodes to Rm State.

### PR DESCRIPTION
**Pull request by tobwiens**
Add maxNumberOfNodes to Rm State.

The maxNumberOfNodes value is added to the RM State. That allows the RM portal to receive the currently set alive nodes limit.

### CODE REVIEW

**Comment by tobwiens** [comment]
Nodes limited to 2 : 



Nodes unlimited (null or anything smaller than zero):

**Comment by tobwiens** [comment]
@ow2-proactive/developers two reviews please.

File: rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/RMStateNodeUrls.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ *
+ *  Initial developer(s):               The ProActive Team
+ *                        http://proactive.inria.fr/team_members.htm
+ *  Contributor(s): ActiveEon Team - http://www.activeeon.com
+ *
+ *  * $$ACTIVEEON_CONTRIBUTOR$$
+ */
+package org.ow2.proactive.resourcemanager.common;
+
+import org.objectweb.proactive.annotation.PublicAPI;
```
Comment (sonarqube-activeeon) [inline rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/RMStateNodeUrls.java]: MINOR Remove this unused import 'org.objectweb.proactive.annotation.PublicAPI'. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/RMStateNodeUrls.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+import javax.xml.bind.annotation.XmlRootElement;
```
Comment (sonarqube-activeeon) [inline rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/RMStateNodeUrls.java]: MINOR Remove this unused import 'javax.xml.bind.annotation.XmlRootElement'. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by sonarqube-activeeon** [comment]
SonarQube analysis reported 4 issues
* MINOR 4 minor

Watch the comments in this conversation to review them.

#### 2 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [RMState.java#L39](https://github.com/ow2-proactive/scheduling/blob/29e3becdea1965da8597c14e5a8c1ee14b30aba2/rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/RMState.java#L39): Remove this unused import 'org.objectweb.proactive.annotation.PublicAPI'. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AUselessImportCheck)
1. MINOR [RMState.java#L42](https://github.com/ow2-proactive/scheduling/blob/29e3becdea1965da8597c14e5a8c1ee14b30aba2/rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/RMState.java#L42): Remove this unused import 'javax.xml.bind.annotation.XmlRootElement'. [rule](http://master.ci.activeeon.com:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by fviale** [comment]
:+1:

**Comment by tobwiens** [comment]
@fviale Thank you for your review.
@ow2-proactive/developers one review please.

**Comment by tobwiens** [comment]
Squashed commits.

**Comment by tobwiens** [comment]
@ow2-proactive/developers one review please.

File: rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/RMStateNodeUrls.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import java.util.Set;
+
+public class RMStateNodeUrls implements Serializable {
+
+    private final Set<String> freeNodesUrls;
+    private final Set<String> aliveNodesUrls;
+    private final Set<String> allNodesUrls;
+
+    public RMStateNodeUrls(Set<String> freeNodesUrls, Set<String> aliveNodesUrls, Set<String> allNodesUrls) {
+        this.freeNodesUrls = freeNodesUrls;
```
Comment (marcocast) [inline rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/RMStateNodeUrls.java]: If I got it right this class represents the state of the RM nodes at certain point in time. If so, the sets should be immutable. For example :
`this.freeNodesUrls = new ImmutableSet.Builder().addAll(freeNodesUrls).build();`

Comment (tobwiens) [inline rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/RMStateNodeUrls.java]: I went down that path, but the scheduling loop relies the mutability of an RM state instance as an example here a line of code: https://github.com/ow2-proactive/scheduling/blob/master/scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingMethodImpl.java#L230

I figured that it is way out of the scope to change the scheduling loop especially knowing the severity if I introduce bugs. 

So I kept it mutable.

Comment (marcocast) [inline rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/RMStateNodeUrls.java]: I see what you mean, it is heavily based on the mutability of the freeResources collections and it's risky to change.
We could have still kept it immutable by returning a copy of the collection : 

`return new HashSet(freeNodesUrls);`

But i see that the SchedulingMethodImpl is adding nodes to the freeResources in many points. 
Looks like to make it immutable is required a broader scope than the scope of this change as you said, so ok.

**Pull request merged**